File: 1010. Pairs of Songs With Total Durations Divisible by 60/1010.py
class Solution:
    def numPairsDivisibleBy60(self, time) -> int:
        myDict = {}
        for i in time:
            v = myDict.get(i%60, 0)
            myDict[i%60] = v+1
        
        count = 0
        
        count += myDict.get(0, 0)*(myDict.get(0, 0)-1)//2
        count += myDict.get(30, 0)*(myDict.get(30, 0)-1)//2
        for i in range(1,30):
            count += myDict.get(i, 0)*myDict.get(60-i, 0)

        return count

    # 想法是类似的不过我用的是字典，这个用的是列表
    def numPairsDivisibleBy601(self, time):
        length = [0 for _ in range(60)]
        for t in time:
            length[t % 60] += 1

        ans = 0

        ans += length[0] * (length[0] - 1) // 2  # 0
        ans += length[30] * (length[30] - 1) // 2  # 30
        for idx in range(1, 30):
			# 1 to 29
            ans += length[idx] * length[60-idx]

        return ans

a = Solution()
print(a.numPairsDivisibleBy60([60,60,60]))

File: 1010. Pairs of Songs With Total Durations Divisible by 60/1010.go
package main

import(
	"fmt"
)

func main(){
	fmt.Println(numPairsDivisibleBy60([]int{60, 60, 60}))
}

func numPairsDivisibleBy60(time []int) int {
	myMap := make(map[int]int)
	for _,v := range time{
		mapv, ok := myMap[v%60]
		if !ok{
			myMap[v%60]=1
		}else{
			myMap[v%60]=mapv+1
		}
	}

	var count int

	v, ok := myMap[0]
	if ok{
		count += v*(v-1)/2
	}
	v, ok = myMap[30]
	if ok{
		count += v*(v-1)/2
	}

	for i:=1; i<30; i++{
		v1, ok1 := myMap[i]
		v2, ok2 := myMap[60-i]
		if ok1 && ok2{
			count += v1*v2
		}
	}

	return count
}

// 很厉害，好想法
func numPairsDivisibleBy601(time []int) int {
    var (
        count int
        record = make([]int, 60)
    )
    for i := 0; i < len(time); i++ {
        mod := time[i] % 60
        if mod == 0 {
            count += record[0] 
        } else {
            count += record[60-mod]
        }        
        record[mod]++
    }
    return count
}